pkg/prometheus/promcfg.go

```diff
@@ -1142,7 +1145,12 @@ func (cg *ConfigGenerator) generateServiceMonitorConfig(
 	}
 
 	selectedNamespaces := getNamespacesFromNamespaceSelector(&m.Spec.NamespaceSelector, m.Namespace, ignoreNamespaceSelectors)
-	cfg = append(cfg, cg.generateK8SSDConfig(selectedNamespaces, apiserverConfig, store, kubernetesSDRoleEndpoint))
+
+	role := kubernetesSDRoleEndpoint
+	if cg.version.GTE(semver.MustParse("2.21.0")) && cg.endpointSliceSupported {
+		role = kubernetesSDRoleEndpointSlice
+	}
```

Comment:
Part of it could be moved to the ConfigGenerator struct for reuse. Something like:

```golang
func (cg *ConfigGenerator) EndpointSliceSupported() bool {
    return cg.version.GTE(semver.MustParse("2.21.0")) && cg.endpointSliceSupported 
}
```